### PR TITLE
first_commit

### PR DESCRIPTION
**Pull request by AntonM455**
first itteration

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -1,40 +1,109 @@
...
-	// ваш код здесь
+// generateRandomElements generates random elements and is recorded in the created slides.
+func generateRandomElements(size int) ([]int, error) {
+	if size <= 0 {
+		return nil, errors.New("size must be greater than 0")
+	}
+
+	data := make([]int, size)
+	for i := range data {
+		data[i] = rand.Intn(1_000_000) // 0 - 999 999
```
Comment (dirtymew) [inline main.go]: в задании нет ничего про ограничение 0 - 999 999

так что нужно чтобы было от (0, 9223372036854775807]

File: main.go [hunk 2]
```diff
@@ -1,40 +1,109 @@
...
 const (
 	SIZE   = 100_000_000
 	CHUNKS = 8
 )
 
-// generateRandomElements generates random elements.
-func generateRandomElements(size int) []int {
-	// ваш код здесь
+// generateRandomElements generates random elements and is recorded in the created slides.
+func generateRandomElements(size int) ([]int, error) {
```
Comment (dirtymew) [inline main.go]: не нужно менять интерфейс функции

Comment (dirtymew) [reply main.go]: func generateRandomElements(size int) []int {

File: main.go [hunk 3]
```diff
@@ -1,40 +1,109 @@
...
+	for i := range data {
+		data[i] = rand.Intn(1_000_000) // 0 - 999 999
+	}
+	return data, nil
 }
 
 // maximum returns the maximum number of elements.
-func maximum(data []int) int {
-	// ваш код здесь
+func maximum(data []int) (int, error) {
```
Comment (dirtymew) [inline main.go]: func maximum(data []int) int {

Comment (dirtymew) [reply main.go]: это учебное задания и не про ошибки 

не меняй его пожалуйста

File: main.go [hunk 4]
```diff
@@ -1,40 +1,109 @@
...
+		start := i * chunkSize
+		end := start + chunkSize
+		if i == CHUNKS-1 {
+			end = len(data)
+		}
+		// launching goroutines
+		go func(index int, chunk []int) {
+			defer wg.Done()
+			localMax, _ := maximum(chunk)
+			mu.Lock()
```
Comment (dirtymew) [inline main.go]: не нужен тут мютекс

каждая горутина работает со своим элементом массива

File: main_test.go [hunk 5]
```diff
@@ -1,3 +1,62 @@
 package main
 
-// Пишите тесты в этом файле
+import (
+	"testing"
+)
+
+// TestGenerateRandomElements checks what happens if you pass 0 elements and what will happen if I transfer 10 elements normally
+func TestGenerateRandomElements(t *testing.T) {
+	_, err := generateRandomElements(0)
```
Comment (dirtymew) [inline main_test.go]: лучше табличные тесты

и обязательно граничные условия проверяй

Comment (dirtymew) [reply main_test.go]: это замечание ко всем тестам

**Review by dirtymew — CHANGES_REQUESTED** [review]
Привет Антон

хорошая работа - но есть что доделать и поправить

File: main.go (hunk 2 above)
Comment (AntonM455) [reply main.go]: Привет! Уберу сейчас тут и везде, уберу обработку ошибок и из возврата функции ошибку

File: main.go (hunk 4 above)
Comment (AntonM455) [reply main.go]: Понял ошибку, сейчас поправлю

File: main_test.go [hunk 6]
```diff
@@ -1,3 +1,74 @@
...
+}
+
+// TestMaximum according to comments, corrected: made a table test and edge cases
+func TestMaximum(t *testing.T) {
+	tests := []struct {
+		name     string
+		data     []int
+		expected int
+	}{
+		{"One element", []int{42}, 42},
```
Comment (dirtymew) [inline main_test.go:37]: еще с nil кейс стоило добавить

Comment (dirtymew) [reply main_test.go:37]: и с пустым слайсом

File: main.go [hunk 7]
```diff
@@ -2,39 +2,81 @@ package main
...
+		data[i] = rand.Intn(math.MaxInt64)
+	}
+	return data
 }
 
 // maximum returns the maximum number of elements.
 func maximum(data []int) int {
-	// ваш код здесь
+	max := data[0]
+	for _, v := range data[1:] {
```
Comment (dirtymew) [inline main.go:32]: с пустым слайсом будут проблемы

File: main.go [hunk 8]
```diff
@@ -2,39 +2,81 @@ package main
...
+	for i := 0; i < CHUNKS; i++ {
+		start := i * chunkSize
+		end := start + chunkSize
+		if i == CHUNKS-1 {
+			end = len(data)
+		}
+		// launching goroutines
+		go func(index int, chunk []int) {
+			defer wg.Done()
+			localMax := maximum(chunk)
```
Comment (dirtymew) [inline main.go:56]: maxSlice[index] = maximum(chunk)

**Review by dirtymew — APPROVED** [review]
Привет Антон

работа принята - но есть что улучшить
